jdat_notebooks/ifu_optimal/ifu_optimal.ipynb

```diff
@@ -0,0 +1,821 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#NIRSpec IFU science data cube\n",
+    "BoxPath=\"https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/IFU_optimal_extraction/\"\n",
+    "filename=BoxPath+'NRS00001-faintQSO-F100LP-G140H-01_1_491_SE_2020-08-25T12h15m00_s3d.fits'\n",
+    "\n",
+    "#Open and inspect the file and WCS\n",
+    "#Load with astropy.fits.open\n",
+    "with fits.open(filename, memmap=False) as hdulist:\n",
+    "    sci = hdulist[\"SCI\"].data\n",
+    "    err = hdulist[\"ERR\"].data\n",
+    "    w = wcs.WCS(hdulist[1].header)\n",
+    "    hdr=hdulist[1].header\n",
+    "    hdulist.info()\n",
+    "    print(w)\n",
+    "    \n",
+    "#Load with Spectrum1D    \n",
+    "spec1d = Spectrum1D.read(filename)\n",
+    "\n",
+    "#Wavelengths\n",
+    "wavelength=np.array(spec1d.spectral_axis.value)\n",
+    "print(wavelength)\n",
+    "\n",
+    "#Sum over spaxels\n",
+    "fnu_sum=np.sum(spec1d.flux,axis=(0,1))\n",
+    "\n",
+    "#List of cube slices\n",
+    "data=[]\n",
+    "var=[]\n",
+    "spec1d_len=len(spec1d.spectral_axis.value)\n",
+    "for idx in range(spec1d_len): \n",
+    "    data.append(sci[idx,:,:])\n",
+    "    var.append(err[idx,:,:])   #variance=err, not variance=err**2.  Squaring the err gives strange, noisy results. \n",
+    "\n",
+    "#Window data, variance, and PSF model (and replace NaNs)\n",
+    "#The existing JWST pipeline window is overgenerous (39x33 instead of the nominal 30x30 pixels)\n",
+    "data_win=np.nan_to_num(np.array(data)[:,5:-4,3:])\n",
+    "data_var=np.nan_to_num(np.array(var)[:,5:-4,3:]) \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "*Developer Note:* Can we fix or suppress this warning?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "*Developer Note:*  Is there a way to read in with only Spectrum1D to perform all of our cube operations, not using fits.open()?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualize Science Data with Cubeviz"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "*Developer Note:* Cubeviz incompatible with jupyter_client 6.1.6.  Use jupyter_client 5.3.5 instead."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "from jdaviz import CubeViz\n",
+    "cubeviz = CubeViz()\n",
+    "cubeviz.app"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### UI Instructions:\n",
+    "* Load science datacube into Cubeviz using the cell below\n",
+    "* Select a circular subset region centered on the source. \n",
```

Comment:
The source is very faint and I could barely see it. I think it shows the highest flux around slice 1607. I was unable to change the contrast in the viewer.

I was unable to zoom into the spectrum. "Select a range of x values" did not work for me even though I tried several key combinations.

Reply:
We should discuss whether the inability to zoom was a glitch, a bug, or something else.  This appears to work for me.
Ditto for changing contrast in the viewer.

The source is very faint.  Maybe I can figure out a Min/Max setting in the viewer and recommend that setting in the markup cell.

Reply:
I find that using Linear stretch at 90 percentile shows the source and have added instructions to that effect